### PR TITLE
Redirects Plugin - Make Relative or Absolute.

### PR DESCRIPTION
**Pull request by tonypartridge**
The redirects plugin always stores the absolute URL. This change adds a config option into the plugin to include the domain or not. 

This means redirects are no longer domain dependant since we added relative redirect support a few versions back.

Config option set to include url by default for BC.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.plg_system_redirect.ini [hunk 1]
```diff
@@ -14,3 +14,5 @@ PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_REGEXP_LABEL="Regular Expression"
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_DESC="A regular expression or a term which should be excluded."
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_LABEL="Term"
 PLG_SYSTEM_REDIRECT_XML_DESCRIPTION="The system redirect plugin enables the Joomla Redirect system to catch missing pages and redirect users."
+PLG_SYSTEM_REDIRECT_FIELD_STORE_FULL_URL_LABEL="Include domain name in ExpiredURL"
+PLG_SYSTEM_REDIRECT_FIELD_STORE_FULL_URL_DESC="When we save the expired url should we include the domain name? if not we will store it relatively i.e. https://www.joomla.org/article becomes /article."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_redirect.ini]: after a ? it is a new sentence so it must start with a capital letter

File: administrator/language/en-GB/en-GB.plg_system_redirect.ini [hunk 2]
```diff
@@ -14,3 +14,5 @@ PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_REGEXP_LABEL="Regular Expression"
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_DESC="A regular expression or a term which should be excluded."
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_LABEL="Term"
 PLG_SYSTEM_REDIRECT_XML_DESCRIPTION="The system redirect plugin enables the Joomla Redirect system to catch missing pages and redirect users."
+PLG_SYSTEM_REDIRECT_FIELD_STORE_FULL_URL_LABEL="Include domain name in ExpiredURL"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_redirect.ini]: add a space in ExpiredURL please

File: administrator/language/en-GB/en-GB.plg_system_redirect.ini [hunk 3]
```diff
@@ -13,4 +13,6 @@ PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_REGEXP_DESC="Should the term be handled a
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_REGEXP_LABEL="Regular Expression"
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_DESC="A regular expression or a term which should be excluded."
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_LABEL="Term"
+PLG_SYSTEM_REDIRECT_FIELD_STORE_FULL_URL_DESC="When we save the expired url should we include the domain name? If not we will store it relatively i.e. https://www.joomla.org/article becomes /article."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_system_redirect.ini]: How about this to keep it short and simple without the references to `we`? 
`Save the expired URL as absolute (include domain) or relative (exclude domain).`

**Comment by tonypartridge** [comment]
should we be keeping it short when providing a description? But I’m easy either way. Thoughts @brianteeman?

**Comment by laoneo** [comment]
I was wondering why the full url is stored at all? If it is not needed, then I would remove it and not make it configurable.

**Comment by tonypartridge** [comment]
It is needed @allon, some sites have multiple domains and https/http. Regardless if it’s right or wrong J3 does this 😬 and must continue so. There was a bit of backlash because my lag change forced lower case url redirects but some people were using upper and lowercase characters in random places....

J4 could go completely relative though?

**Comment by infograf768** [comment]
Note:
When doing https://github.com/joomla/joomla-cms/pull/19734 I remarked that it worked perfectly when the oldurls were created by the plugin, but NOT when creating them in the component manager.
Whatever the decision to include or not the absolute url, would be good to have a look at the fact that manual entries are not checked the same way as automatic.

**Comment by tonypartridge** [comment]
How do you mean? For this pr we wouldn’t want to prevent the url in Manual entries given we still support urls it’s just this config sets the plugin to auto store in relative or absolute.

Or do you mean when adding manually they are not stored utf8? If so that’s an easy enough fix.

**Comment by infograf768** [comment]
> Or do you mean when adding manually they are not stored utf8? If so that’s an easy enough fix.

They are stored as utf8, but we do not have the check for duplicates that we get for the urls entered automatically by the plugin.

**Comment by tonypartridge** [comment]
@infograf768 at which point is this check needed? When we add a url it's checked if it exists before adding it? If it doesn't exist it isn't added.. if it exists it's updated..

Am I completely missing your point here?

**Comment by infograf768** [comment]
The check is not done when creating a new Redirect.
Let me make a gif.

**Comment by infograf768** [comment]
redirect

**Comment by tonypartridge** [comment]
@infograf768 thanks! testing / checking now.

**Comment by tonypartridge** [comment]
This was broken in 3.8.6 it worked in 3.8.5. Looking into it now.

**Comment by infograf768** [comment]
As long as it does not break https://github.com/joomla/joomla-cms/pull/19734 😃

**Comment by tonypartridge** [comment]
The problem is @infograf768 your pr breaks checks to validate against duplicate urls since it encodes them... on the stored URL. But the non-stored URL is not stored as rawurlencode. 

Looking at a correct fix now

**Comment by infograf768** [comment]
Any working solution which also solves the original issue "solved" by #19734 is welcome

**Comment by tonypartridge** [comment]
@infograf768 I've just tested without your PR and both:

http://mydomain.com/bogusurl
http://mydomain.com/bøgusurl

are treated separately. The problem with your PR is the check against the DB is never valid since we do not store as rawurlencode. It actually just breaks the check completely. 

Are you saying both these links should be treated separately or equally?

**Comment by infograf768** [comment]
> Are you saying both these links should be treated separately or equally?

They should be treated separately. That was the whole goal of my PR.
See the original issue
https://github.com/joomla/joomla-cms/issues/19713

specially that part
> On save it will give the error "Save failed with the following error: The source URL must be unique."
Test the URL and it will give you a 404 page. It will not use the redirect set in step No. 4.

**Comment by infograf768** [comment]
Basically, the check DOES work fine when the db is already populated by the plugin, but not when the redirect is fully created in the manager.

**Comment by tonypartridge** [comment]
I’ve done a pr to revert your pr and will test throughly the difference between plugin and component created urls.

**Comment by infograf768** [comment]
If you don't mind, a PR to revert my PR is not the solution as 99% of needed redirects are done by the plugin and my PR works fine for these.

**Comment by tonypartridge** [comment]
@infograf768, so I reverted your PR and ran some tests.

I enabled the plugin and accessed:
https://mywebsite.com/bøgusurl-auto

404 adding to com_redirects as:
https://mywebsite.com/bøgusurl-auto

I then did:
https://mywebsite.com/bogusurl-auto

404 adding to com_redirects as:
https://mywebsite.com/bogusurl-auto

Both redirect independently. 

I also then tested creating manual entries, again both validations are unique, I cannot add https://mywebsite.com/bogusurl-auto if it exists and can add https://mywebsite.com/bogusurl-auto if https://mywebsite.com/bøgusurl-auto exists and https://mywebsite.com/bogusurl-auto doesn't. 

Can you please test on your site with the PR reverted as pr my other pr. And provide a gif of the issue you are facing?

Many thanks

File: administrator/language/en-GB/en-GB.plg_system_redirect.ini (hunk 3 above)
Comment (tonypartridge) [reply administrator/language/en-GB/en-GB.plg_system_redirect.ini]: @nibra liked and you suggested so I've applied it :-) Thanks!

**Comment by brianteeman** [comment]
Unless you have some hidden tracking or a crystal ball you cannot say this.

 > 99% of needed redirects

**Comment by infograf768** [comment]
@tonypartridge 
siesta now. will post an animated gif of the original issue later on.

**Comment by infograf768** [comment]
@tonypartridge 

This is the issue before my PR. To test:
1. Revert https://github.com/joomla/joomla-cms/pull/19734
2. In frontend, enter a `bogusurl`,  get 404, edit it in back-end, then try a `bôgusurl`, get a 404, edit it in backend, try to save.

This is what we get before my PR.
redirect_2

After my PR, we can save the `bôgusurl` expired url with a New URL.

**Comment by tonypartridge** [comment]
now replicated thank you! Looking into it.

**Comment by tonypartridge** [comment]
@infograf768  see #19950 for your UTF8 support. 

Can we please get some tests on this PR in the mean time :-)

**Comment by infograf768** [comment]
#19950 works great! Needs another tester.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 2499d3e7ba0009304dbc34716144c79c5cfc98b1<br /><br />Tested OK here.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19942">issues.joomla.org/tracker/joomla-cms/19942</a>.</sub>

File: administrator/language/en-GB/en-GB.plg_system_redirect.ini [hunk 4]
```diff
@@ -13,4 +13,6 @@ PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_REGEXP_DESC="Should the term be handled a
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_REGEXP_LABEL="Regular Expression"
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_DESC="A regular expression or a term which should be excluded."
 PLG_SYSTEM_REDIRECT_FIELD_EXCLUDE_URLS_TERM_LABEL="Term"
+PLG_SYSTEM_REDIRECT_FIELD_STORE_FULL_URL_DESC="Save the expired URL as absolute (include domain) or relative (exclude domain)."
+PLG_SYSTEM_REDIRECT_FIELD_STORE_FULL_URL_LABEL="Include domain name in Expired URL"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_system_redirect.ini]: Change `domain name` to `Domain Name` since labels are capitalized.

**Comment by tonypartridge** [comment]
Done

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on de7219808a48262aedf106df7b8bbd6d0ad9efd8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19942">issues.joomla.org/tracker/joomla-cms/19942</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19942">issues.joomla.org/tracker/joomla-cms/19942</a>.</sub>

**Pull request merged**